File: 338.-Counting-Bits/counting_bits_v2.cpp
#include <iostream>
#include <vector>

using namespace std;

vector <int> countBits(int num) {
	vector <int> result(num + 1, 0);
	unsigned int base_pow = 1, flag = 0;

	for (unsigned int i = 1; i < result.size(); i++)
	{
		if (i == base_pow)
		{
			base_pow *= 2;
			flag = 0;
		}
		result[i] = result[flag] + 1;
		flag++;
	}
	return result;
}

int main(int argc, char const *argv[])
{
	vector <int> test;
	test = countBits(5);
	cout << "[";
	for (unsigned int i = 0; i < test.size() - 1; i++)
	{
		cout << test[i] << " ";
	}
	cout << test[test.size() - 1];
	cout << "]" << endl;
	return 0;
}

File: 338. Counting Bits/counting_bits_v2.py
class Solution(object):

    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
        list1 = []
        base_pow = 1
        flag = 0
        for n in range (0, num + 1):
            list1.append(0)

        for n in range(1, num + 1):
            if n == base_pow:
                base_pow = base_pow * 2
                flag = 0
            list1[n] = list1[flag] + 1
            flag = flag + 1

        return list1
